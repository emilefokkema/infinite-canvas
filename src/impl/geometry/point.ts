export class Point{
	constructor(public readonly x: number, public readonly y: number){}
	public mod(): number{
		return Math.sqrt(this.modSq());
	}
	public modSq(): number{
		return this.x * this.x + this.y * this.y;
	}
	public minus(other: Point): Point{
		return new Point(this.x - other.x, this.y - other.y);
	}
	public plus(other: Point): Point{
		return new Point(this.x + other.x, this.y + other.y);
	}
	public dot(other: Point): number{
		return this.x * other.x + this.y * other.y;
	}
	public cross(other: Point): number{
		return this.x * other.y - this.y * other.x;
	}
	public equals(other: Point): boolean{
		return this.x === other.x && this.y === other.y;
	}
	public getPerpendicular(): Point{
		return new Point(-this.y, this.x);
	}
	public scale(r: number): Point{
		return new Point(r * this.x, r * this.y);
	}
	public projectOn(other: Point): Point{
		return other.scale(this.dot(other) / other.modSq());
	}
	public matrix(a: number, b: number, c: number, d: number): Point{
		return new Point(a * this.x + b * this.y, c * this.x + d * this.y);
	}
	public inSameDirectionAs(other: Point): boolean{
		return this.cross(other) === 0 && this.dot(other) >= 0;
	}
	public isInOppositeDirectionAs(other: Point): boolean{
		return this.cross(other) === 0 && this.dot(other) < 0;
	}
	public isOnSameSideOfOriginAs(other1: Point, other2: Point){
		return this.isInSmallerAngleBetweenPoints(other1, other2) || 
			   other1.isInSmallerAngleBetweenPoints(this, other2) ||
			   other2.isInSmallerAngleBetweenPoints(this, other1);
	}
	public isInSmallerAngleBetweenPoints(point1: Point, point2: Point): boolean{
		const cross: number = point1.cross(point2);
		if(cross > 0){
			return point1.cross(this) >= 0 && this.cross(point2) >= 0;
		}else if(cross < 0){
			return point1.cross(this) <= 0 && this.cross(point2) <= 0;
		}else if(point1.dot(point2) > 0){
			return this.cross(point1) === 0 && this.dot(point1) > 0;
		}else{
			return true;
		}
	}
	public static origin: Point = new Point(0, 0);
}