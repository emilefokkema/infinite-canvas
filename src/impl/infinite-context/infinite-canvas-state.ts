import { ViewBox } from "../interfaces/viewbox";

export class InfiniteCanvasState implements CanvasState{
	constructor(private viewBox: ViewBox){}
	public restore():void{
		this.viewBox.restoreState();
	}
	public save(): void{
		this.viewBox.saveState();
	}
	public reset(): void{
		this.viewBox.resetState();
	}
}