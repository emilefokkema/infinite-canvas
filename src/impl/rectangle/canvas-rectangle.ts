import { ConvexPolygon } from "../areas/polygons/convex-polygon";
import { Point } from "../geometry/point";
import { CoordinateSystem } from "./coordinate-system";
import { CanvasMeasurement } from "./canvas-measurement";
import { Transformation } from "../transformation";
import { TransformationRepresentation } from "api/transformation-representation";
import { Units } from "api/units";

export interface CanvasRectangle{
    readonly viewboxWidth: number;
    readonly viewboxHeight: number;
    readonly polygon: ConvexPolygon;
    readonly userTransformation: Transformation;
    readonly infiniteCanvasContext: CoordinateSystem;
    readonly initialBitmapTransformation: Transformation;
    addPathAroundViewbox(context: CanvasRenderingContext2D, margin: number, counterclockwise: boolean): void;
    getCSSPosition(clientX: number, clientY: number): Point;
    getTransformationForInstruction(infiniteCanvasContextTransformation: TransformationRepresentation): Transformation;
    translateInfiniteCanvasContextTransformationToBitmapTransformation(infiniteCanvasContextTransformation: TransformationRepresentation): Transformation;
    getBitmapTransformationToTransformedInfiniteCanvasContext(): Transformation;
    getBitmapTransformationToInfiniteCanvasContext(): Transformation;
    withMeasurement(measurement: CanvasMeasurement): CanvasRectangle
    withTransformation(transformation: TransformationRepresentation): CanvasRectangle
    withUnits(units: Units): CanvasRectangle
}