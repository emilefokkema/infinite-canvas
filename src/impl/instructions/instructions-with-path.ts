import { StateChangingInstructionSequence } from "./state-changing-instruction-sequence";
import { PreExecutableInstructionWithState } from "./pre-executable-instruction-with-state";
import { CurrentPath } from "../interfaces/current-path";
import { PathInstruction } from "../interfaces/path-instruction";
import { InfiniteCanvasState } from "../state/infinite-canvas-state";
import { Instruction } from "./instruction";
import { Area } from "../areas/area";
import { InfiniteCanvasAreaBuilder } from "../areas/infinite-canvas-area-builder";
import { Position } from "../geometry/position";
import { transformPosition } from "../geometry/transform-position";
import { InstructionsWithSubpath } from "./instructions-with-subpath";
import { DrawnPathProperties } from "../interfaces/drawn-path-properties";
import { InfiniteCanvasPathInfinityProvider } from "../infinite-canvas-path-infinity-provider";
import { InstructionsToClip } from "../interfaces/instructions-to-clip";
import { InstructionsToClipImpl } from "./instructions-to-clip-impl";
import { ExecutableStateChangingInstructionSet } from "../interfaces/executable-state-changing-instruction-set";
import { ExecutableStateChangingInstructionSequence } from "./executable-state-changing-instruction-sequence";
import { ExecutableInstructionWithState } from "./executable-instruction-with-state";

export class InstructionsWithPath extends StateChangingInstructionSequence<InstructionsWithSubpath> implements CurrentPath{
    private areaBuilder: InfiniteCanvasAreaBuilder = new InfiniteCanvasAreaBuilder();
    constructor(private _initiallyWithState: PreExecutableInstructionWithState){
        super(_initiallyWithState);
    }
    public get area(): Area{return this.areaBuilder.area;}
    public surroundsFinitePoint(): boolean{
        for(const subpath of this.added){
            if(subpath.surroundsFinitePoint()){
                return true;
            }
        }
        return false;
    }
    public currentSubpathIsClosable(): boolean{
        if(this.added.length === 0){
            return true;
        }
        return this.added[this.added.length - 1].isClosable();
    }
    public allSubpathsAreClosable(): boolean{
        if(this.added.length === 0){
            return true;
        }
        for(const subPath of this.added){
            if(!subPath.isClosable()){
                return false;
            }
        }
        return true;
    }
    public drawPath(instruction: Instruction, state: InfiniteCanvasState, drawnPathProperties: DrawnPathProperties): ExecutableStateChangingInstructionSet{
        if(this.added.length === 0){
            return;
        }
        const toAdd = ExecutableInstructionWithState.create(state, instruction);
        const result = this.makeExecutable(drawnPathProperties);
        toAdd.setInitialState(result.state)
        result.add(toAdd)
        return result;
    }
    private makeExecutable(drawnPathProperties: DrawnPathProperties): ExecutableStateChangingInstructionSequence<ExecutableStateChangingInstructionSet>{
        const result = new ExecutableStateChangingInstructionSequence<ExecutableStateChangingInstructionSet>(this._initiallyWithState.makeExecutable());
        const pathInfinityProvider = new InfiniteCanvasPathInfinityProvider(drawnPathProperties)
        for(const added of this.added){
            result.add(added.makeExecutable(pathInfinityProvider))
        }
        return result;
    }
    private getInstructionsToClip(): InstructionsToClip{
        const executable = this.makeExecutable({lineWidth: 0, lineDashPeriod: 0, shadowOffsets: []})
        executable.setInitialState(executable.stateOfFirstInstruction)
        return new InstructionsToClipImpl(executable, this.area)
    }
    public clipPath(instruction: Instruction, state: InfiniteCanvasState): void{
        if(this.added.length === 0){
            return;
        }
        const currentSubpath: InstructionsWithSubpath = this.added[this.added.length - 1];
        const toAdd: PreExecutableInstructionWithState = PreExecutableInstructionWithState.create(state, instruction);
        currentSubpath.addInstruction(toAdd);
        const instructionsToClip = this.getInstructionsToClip();
        this.addClippedPath(instructionsToClip);
    }
    public closePath(): void{
        if(this.added.length === 0){
            return;
        }
        const currentSubpath: InstructionsWithSubpath = this.added[this.added.length - 1];
        currentSubpath.closePath();
    }
    public moveTo(position: Position, state: InfiniteCanvasState): void{
        const transformedPointToMoveTo: Position = transformPosition(position, state.current.transformation);
        this.areaBuilder.addPosition(transformedPointToMoveTo);
        const newSubpath: InstructionsWithSubpath = InstructionsWithSubpath.create(state, position);
        newSubpath.setInitialState(this.state);
        this.add(newSubpath);
    }
    public canAddLineTo(position: Position, state: InfiniteCanvasState): boolean{
        if(this.added.length === 0){
          return true;
        }
        const transformedPosition: Position = transformPosition(position, state.current.transformation);
        return this.added[this.added.length - 1].canAddLineTo(transformedPosition);
    }
    public lineTo(position: Position, state: InfiniteCanvasState): void{
        if(this.added.length === 0){
            this.moveTo(position, state);
            return;
        }
        const currentSubpath: InstructionsWithSubpath = this.added[this.added.length - 1];
        const transformedPosition: Position = transformPosition(position, state.current.transformation);
        this.areaBuilder.addPosition(transformedPosition);
        currentSubpath.lineTo(position, state);
    }
    public addPathInstruction(pathInstruction: PathInstruction, state: InfiniteCanvasState): void{
        if(this.added.length === 0){
            if(pathInstruction.initialPoint){
                this.moveTo(pathInstruction.initialPoint, state);
            }else{
                return;
            }
        }
        const currentSubpath: InstructionsWithSubpath = this.added[this.added.length - 1];
        const currentPosition: Position = currentSubpath.currentPosition;
        const inverseTransformedCurrentPosition = transformPosition(currentPosition, state.current.transformation.inverse());
        pathInstruction.changeArea(this.areaBuilder.transformedWith(state.current.transformation), inverseTransformedCurrentPosition);
        const toAdd: PreExecutableInstructionWithState = PreExecutableInstructionWithState.create(state, pathInstruction.instruction);
        currentSubpath.addPathInstruction(pathInstruction, toAdd, state);
    }
    public static create(initialState: InfiniteCanvasState): InstructionsWithPath{
        return new InstructionsWithPath(PreExecutableInstructionWithState.create(initialState, (context: CanvasRenderingContext2D) => {context.beginPath();}));
    }
}
