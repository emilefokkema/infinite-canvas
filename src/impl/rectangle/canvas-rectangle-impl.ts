import { CanvasRectangle } from './canvas-rectangle';
import { CoordinatesSwitch } from './coordinates-switch';
import { CanvasMeasurement } from "./canvas-measurement";
import { ConvexPolygon } from "../areas/polygons/convex-polygon";
import { Point } from "../geometry/point";
import { TransformationRepresentation } from "api/transformation-representation";
import { Transformation } from "../transformation";
import { Units } from "api/units";
import { CoordinateSystem } from './coordinate-system';
import { getRectWithFourEdges } from '../rect/get-rect-strategy';

function measurementsAreOfEqualSize(one: CanvasMeasurement, other: CanvasMeasurement): boolean{
    if(!one){
        return !other;
    }
    if(!other){
        return false;
    }
    return one.viewboxWidth === other.viewboxWidth &&
            one.viewboxHeight === other.viewboxHeight &&
            one.screenWidth === other.screenWidth &&
            one.screenHeight === other.screenHeight;
}

export class CanvasRectangleImpl implements CanvasRectangle{
    public get viewboxWidth(): number{return this.measurement.viewboxWidth;}
    public get viewboxHeight(): number{return this.measurement.viewboxHeight}
    public get userTransformation(): Transformation{return this.coordinates.userTransformation;}
    public get infiniteCanvasContext(): CoordinateSystem{return this.coordinates.infiniteCanvasContext;}
    public get initialBitmapTransformation(): Transformation{return this.coordinates.initialBitmapTransformation;}
    constructor(
        private readonly coordinates: CoordinatesSwitch,
        private readonly measurement: CanvasMeasurement,
        public readonly polygon: ConvexPolygon){
    }

    public withUnits(units: Units): CanvasRectangle{
        const coordinatesSwitch = this.coordinates.withUnits(units);
        return new CanvasRectangleImpl(coordinatesSwitch, this.measurement, this.polygon)
    }

    public withTransformation(transformation: TransformationRepresentation): CanvasRectangle{
        const coordinatesSwitch = this.coordinates.withUserTransformation(Transformation.create(transformation));
        return new CanvasRectangleImpl(coordinatesSwitch, this.measurement, this.polygon);
    }

    public withMeasurement(measurement: CanvasMeasurement): CanvasRectangle{
        const isEqual = measurementsAreOfEqualSize(this.measurement, measurement);
        if(isEqual){
            return this;
        }
        const {viewboxWidth, viewboxHeight} = measurement;
        const polygon = getRectWithFourEdges(0, 0, viewboxWidth, viewboxHeight).getArea();
        const coordinatesSwitch = this.coordinates.withCanvasMeasurement(measurement);
        return new CanvasRectangleImpl(coordinatesSwitch, measurement, polygon)
    }

    public getCSSPosition(clientX: number, clientY: number): Point{
        const {left, top} = this.measurement
        return new Point(clientX - left, clientY - top);
    }

    public getTransformationForInstruction(infiniteCanvasContextTransformation: TransformationRepresentation): Transformation{
        return this.coordinates.getTransformationForInstruction(infiniteCanvasContextTransformation)
    }

    public translateInfiniteCanvasContextTransformationToBitmapTransformation(infiniteCanvasContextTransformation: TransformationRepresentation): Transformation{
        return this.coordinates.translateInfiniteCanvasContextTransformationToBitmapTransformation(infiniteCanvasContextTransformation)
    }

    public getBitmapTransformationToTransformedInfiniteCanvasContext(): Transformation{
        return this.coordinates.getBitmapTransformationToTransformedInfiniteCanvasContext();
    }

    public getBitmapTransformationToInfiniteCanvasContext(): Transformation{
        return this.coordinates.getBitmapTransformationToInfiniteCanvasContext();
    }

    public addPathAroundViewbox(context: CanvasRenderingContext2D, margin: number): void{
        const width: number = this.viewboxWidth + 2 * margin;
        const height: number = this.viewboxHeight + 2 * margin;
        context.save();
        context.setTransform(1, 0, 0, 1, 0, 0);
        context.rect(-margin, -margin, width, height);
        context.restore();
    }

    public static create(measurement: CanvasMeasurement, units: Units): CanvasRectangleImpl{
        const {viewboxWidth, viewboxHeight} = measurement;
        const polygon = getRectWithFourEdges(0, 0, viewboxWidth, viewboxHeight).getArea();
        const coordinates = CoordinatesSwitch.create(units, measurement);
        return new CanvasRectangleImpl(coordinates, measurement, polygon)
    }
}