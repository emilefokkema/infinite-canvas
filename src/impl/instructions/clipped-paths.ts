import { InstructionsToClip } from "../interfaces/instructions-to-clip";
import { Area } from "../areas/area";
import { InfiniteCanvasState } from "../state/infinite-canvas-state";
import { Instruction } from "./instruction";
import { CanvasRectangle } from "../rectangle/canvas-rectangle";

export class ClippedPaths {
    constructor(public area: Area, public latestClippedPath: InstructionsToClip, public readonly previouslyClippedPaths?: ClippedPaths){}
    public withClippedPath(latestClippedPath: InstructionsToClip): ClippedPaths{
        const newArea: Area = latestClippedPath.area.intersectWith(this.area);
        return new ClippedPaths(newArea, latestClippedPath, this);
    }
    public get initialState(): InfiniteCanvasState{
        return this.previouslyClippedPaths ? this.previouslyClippedPaths.initialState : this.latestClippedPath.initialState;
    }
    public except(other: ClippedPaths): ClippedPaths{
        if(other === this){
            return undefined;
        }
        if(this.previouslyClippedPaths){
            return new ClippedPaths(this.area, this.latestClippedPath, this.previouslyClippedPaths.except(other))
        }
        return this;
    }
    public contains(other: ClippedPaths): boolean{
        if(!other){
            return false;
        }
        if(this === other){
            return true;
        }
        if(this.previouslyClippedPaths){
            return this.previouslyClippedPaths.contains(other);
        }
        return false;
    }
    public getInstructionToRecreate(): Instruction{
        const instructionToRecreateLatest: Instruction = (context: CanvasRenderingContext2D, rectangle: CanvasRectangle) => {
            this.latestClippedPath.execute(context, rectangle);
        };
        if(this.previouslyClippedPaths){
            const instructionToRecreatePrevious: Instruction = this.previouslyClippedPaths.getInstructionToRecreate();
            const instructionToConvertToLatest: Instruction = this.previouslyClippedPaths.latestClippedPath.state.getInstructionToConvertToState(this.latestClippedPath.initialState);
            return (context: CanvasRenderingContext2D, rectangle: CanvasRectangle) => {
                instructionToRecreatePrevious(context, rectangle);
                instructionToConvertToLatest(context, rectangle);
                instructionToRecreateLatest(context, rectangle);
            };
        }
        return instructionToRecreateLatest;
    }
}
