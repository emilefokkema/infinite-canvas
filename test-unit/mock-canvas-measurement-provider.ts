import { CanvasMeasurement } from "../src/rectangle/canvas-measurement";
import { CanvasMeasurementProvider } from "../src/rectangle/canvas-measurement-provider";

export class MockCanvasMeasurementProvider implements CanvasMeasurementProvider{
    public measurement: CanvasMeasurement;
    constructor(private readonly width: number, private readonly height: number){
        this.measurement = {
            left: 0,
            top: 0,
            screenWidth: this.width,
            screenHeight: this.height,
            viewboxWidth: this.width,
            viewboxHeight: this.height
        };
    }

    public measure(): CanvasMeasurement{
        return this.measurement;
    }
}
