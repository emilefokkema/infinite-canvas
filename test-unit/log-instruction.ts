import { Instruction } from "../src/instructions/instruction";
import { CanvasContextMock } from "./canvas-context-mock";
import { Transformation } from "../src/transformation";

export function logInstruction(instruction: Instruction, transformation: Transformation = Transformation.identity): string[]{
    const mock: CanvasContextMock = new CanvasContextMock();
    instruction(mock.mock, transformation);
    return mock.getLog();
}