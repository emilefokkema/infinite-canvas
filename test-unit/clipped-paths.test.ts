import { beforeEach, describe, expect, it} from 'vitest'
import { ClippedPaths } from "../src/instructions/clipped-paths";
import { InfiniteCanvasState } from "../src/state/infinite-canvas-state";
import { logInstruction } from "./log-instruction";
import { defaultState } from "../src/state/default-state";
import { fillStyle } from "../src/state/dimensions/fill-stroke-style";
import { InstructionsWithPath } from "../src/instructions/instructions-with-path";
import { Point } from "../src/geometry/point";
import { Instruction } from "../src/instructions/instruction";
import { CanvasRectangle } from "../src/rectangle/canvas-rectangle";
import { CanvasRectangleImpl } from "../src/rectangle/canvas-rectangle-impl";
import { MockCanvasMeasurementProvider } from "./mock-canvas-measurement-provider";

describe("a clipped paths", () => {
    let clippedPaths: ClippedPaths;
    let currentState: InfiniteCanvasState;
    let rectangle: CanvasRectangle;

    beforeEach(() => {
        rectangle = new CanvasRectangleImpl(new MockCanvasMeasurementProvider(200, 200), {});
    });

    describe("and another one", () => {
        let clippedPath: InstructionsWithPath;
        let otherOne: ClippedPaths;

        beforeEach(() => {
            currentState = defaultState;
            clippedPath = InstructionsWithPath.create(currentState, rectangle);
            clippedPath.moveTo(new Point(0, 0), currentState);
            clippedPath.lineTo(new Point(1, 0), currentState);
            clippedPath.lineTo(new Point(1, 1), currentState);
            clippedPath.clipPath((context: CanvasRenderingContext2D) => {context.clip();}, currentState);
            currentState = clippedPath.state;
            clippedPaths = currentState.current.clippedPaths;
            otherOne = clippedPaths;
        });

        describe("containing one more clipped path", () => {

            beforeEach(() => {
                clippedPath.lineTo(new Point(0, 1), currentState);
                clippedPath.clipPath((context: CanvasRenderingContext2D) => {context.clip();}, currentState);
                otherOne = clippedPath.state.current.clippedPaths;
            });

            it("then one should contain the other, but not the other way around", () => {
                expect(otherOne.contains(clippedPaths)).toBe(true);
                expect(clippedPaths.contains(otherOne)).toBe(false);
            });

            describe("and then the other is recreated, starting from the first", () => {
                let recreation: Instruction;

                beforeEach(() => {
                    recreation = otherOne.except(clippedPaths).getInstructionToRecreate(rectangle);
                });

                it("then the recreation should contain the difference", () => {
                    expect(logInstruction(recreation)).toMatchSnapshot();
                });
            });

            describe("and another one, still based on the same path", () => {

                beforeEach(() => {
                    clippedPath.lineTo(new Point(0, 0), currentState);
                    clippedPath.clipPath((context: CanvasRenderingContext2D) => {context.clip();}, currentState);
                    otherOne = clippedPath.state.current.clippedPaths;
                });

                describe("and then the other is recreated, starting from the first", () => {
                    let recreation: Instruction;
    
                    beforeEach(() => {
                        recreation = otherOne.except(clippedPaths).getInstructionToRecreate(rectangle);
                    });
    
                    it("then the recreation should contain the difference", () => {
                        expect(logInstruction(recreation)).toMatchSnapshot();
                    });
                });
            });

            describe("and another one", () => {

                beforeEach(() => {
                    currentState = clippedPath.state;
                    const otherClippedPath: InstructionsWithPath = InstructionsWithPath.create(currentState, rectangle);
                    currentState = currentState.withCurrentState(fillStyle.changeInstanceValue(currentState.current, "#f00"));
                    otherClippedPath.moveTo(new Point(1, 1), currentState);
                    otherClippedPath.clipPath((context: CanvasRenderingContext2D) => {context.clip();}, currentState);
                    otherOne = otherClippedPath.state.current.clippedPaths;
                });

                describe("and then the other is recreated, starting from the first", () => {
                    let recreation: Instruction;
    
                    beforeEach(() => {
                        recreation = otherOne.except(clippedPaths).getInstructionToRecreate(rectangle);
                    });
    
                    it("then the recreation should contain the difference", () => {
                        expect(logInstruction(recreation)).toMatchSnapshot();
                    });
                });
            });
        });
    });
});
