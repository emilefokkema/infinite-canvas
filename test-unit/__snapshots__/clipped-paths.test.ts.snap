// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`a clipped paths > and another one > containing one more clipped path > and another one > and then the other is recreated, starting from the first > then the recreation should contain the difference 1`] = `
[
  "context.beginPath()",
  "context.moveTo(0,0)",
  "context.lineTo(1,0)",
  "context.lineTo(1,1)",
  "context.clip()",
  "context.lineTo(0,1)",
  "context.clip()",
  "context.beginPath()",
  "context.fillStyle = \\"#f00\\"",
  "context.moveTo(1,1)",
  "context.clip()",
]
`;

exports[`a clipped paths > and another one > containing one more clipped path > and another one, still based on the same path > and then the other is recreated, starting from the first > then the recreation should contain the difference 1`] = `
[
  "context.beginPath()",
  "context.moveTo(0,0)",
  "context.lineTo(1,0)",
  "context.lineTo(1,1)",
  "context.clip()",
  "context.lineTo(0,1)",
  "context.clip()",
  "context.lineTo(0,0)",
  "context.clip()",
]
`;

exports[`a clipped paths > and another one > containing one more clipped path > and then the other is recreated, starting from the first > then the recreation should contain the difference 1`] = `
[
  "context.beginPath()",
  "context.moveTo(0,0)",
  "context.lineTo(1,0)",
  "context.lineTo(1,1)",
  "context.clip()",
  "context.lineTo(0,1)",
  "context.clip()",
]
`;
